File: node/src/crypto/ed25519.ts
import * as ed25519 from '@stablelib/ed25519';
import {
    Bytes,
    PublicAlgorithm,
    PublicKey,
    SignaturePrivateKey,
    SignaturePublicKey
} from './types';
import * as proto from '../proto';
import { KeyFormat } from '../proto';

export class Ed25519PrivateKey implements SignaturePrivateKey {
    private readonly key: Uint8Array;

    constructor(
        key: Uint8Array,
    ) {
        switch (key.length) {
            case 32:
                const keyPair = ed25519.generateKeyPairFromSeed(key);
                this.key = keyPair.secretKey;
                break
            case 64:
                this.key = key;
                break
            default:
                throw new Error(`invalid key size: ${key.length}`);
        }
    }

    isDHKey(): boolean {
        return false;
    }

    isSignatureKey(): boolean {
        return true;
    }

    algorithm(): PublicAlgorithm {
        return ed25519Algorithm;
    }

    toPublic(): Ed25519PublicKey {
        return new Ed25519PublicKey(
            ed25519.extractPublicKeyFromSecretKey(this.key),
        )
    }

    getPublic(): PublicKey {
        return this.toPublic();
    }

    getSignaturePublicKey(): SignaturePublicKey {
        return this.toPublic();
    }

    async sign(data: Bytes): Promise<Bytes> {
        return ed25519.sign(this.key, data);
    }
}

export class Ed25519PublicKey implements SignaturePublicKey {
    constructor(public readonly key: Uint8Array) {
    }

    isDHKey(): boolean {
        return false;
    }

    isSignatureKey(): boolean {
        return true;
    }

    algorithm(): PublicAlgorithm {
        return ed25519Algorithm;
    }

    async verify(data: Bytes, signature: Bytes): Promise<boolean> {
        return ed25519.verify(this.key, data, signature);
    }

    marshalToProto(): proto.PublicKey {
        return {
            format: KeyFormat.KeyFormatEd25519,
            data: this.key,
        };
    }

    marshal(): Bytes {
        return this.key;
    }
}

export class Ed25519Algorithm implements PublicAlgorithm {
    getKeyFormat(): proto.KeyFormat {
        return proto.KeyFormat.KeyFormatEd25519;
    }

    async unmarshalPublicKey(input: Uint8Array): Promise<PublicKey> {
        return new Ed25519PublicKey(input);
    }

    generate(): {
        private: Ed25519PrivateKey,
        public: Ed25519PublicKey,
    } {
        const keyPair = ed25519.generateKeyPair();

        return {
            private: new Ed25519PrivateKey(keyPair.secretKey),
            public: new Ed25519PublicKey(keyPair.publicKey)
        };
    }
}

export const ed25519Algorithm = new Ed25519Algorithm();

File: go/cryptoutil/ed25519.go
package cryptoutil

import (
	"crypto/rand"
	"fmt"
	"github.com/jc-lab/jscp/go/payloadpb"
	"golang.org/x/crypto/ed25519"
)

var ed25519Algorithm Ed25519Algorithm

type Ed25519Algorithm struct {
}

func (a *Ed25519Algorithm) GetKeyFormat() payloadpb.KeyFormat {
	return payloadpb.KeyFormat_KeyFormatEd25519
}

func (a *Ed25519Algorithm) UnmarshalPublicKey(input []byte) (PublicKey, error) {
	publicKey := ed25519.PublicKey(input)
	if len(input) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("invalid Key size expected=%d, actual=%d", ed25519.PublicKeySize, len(input))
	}
	instance := &Ed25519PublicKey{
		Key: publicKey,
	}
	return instance, nil
}

func (a *Ed25519Algorithm) Generate() (*Ed25519PrivateKey, error) {
	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	return &Ed25519PrivateKey{
		Key: privateKey,
	}, nil
}

type Ed25519PublicKey struct {
	Key ed25519.PublicKey
}

func (k *Ed25519PublicKey) IsDHKey() bool {
	return false
}

func (k *Ed25519PublicKey) IsSignatureKey() bool {
	return true
}

func (k *Ed25519PublicKey) Algorithm() PublicAlgorithm {
	return &ed25519Algorithm
}

func (k *Ed25519PublicKey) Verify(data []byte, signature []byte) (bool, error) {
	return ed25519.Verify(k.Key, data, signature), nil
}

func (k *Ed25519PublicKey) MarshalToProto() (*payloadpb.PublicKey, error) {
	return &payloadpb.PublicKey{
		Format: k.Algorithm().GetKeyFormat(),
		Data:   k.Key,
	}, nil
}

type Ed25519PrivateKey struct {
	Key ed25519.PrivateKey
}

func (k *Ed25519PrivateKey) IsDHKey() bool {
	return false
}

func (k *Ed25519PrivateKey) IsSignatureKey() bool {
	return true
}

func (k *Ed25519PrivateKey) Algorithm() PublicAlgorithm {
	return &ed25519Algorithm
}

func (k *Ed25519PrivateKey) GetPublic() PublicKey {
	return k.GetSignaturePublicKey()
}

func (k *Ed25519PrivateKey) GetSignaturePublicKey() SignaturePublicKey {
	return &Ed25519PublicKey{
		Key: k.Key.Public().(ed25519.PublicKey),
	}
}

func (k *Ed25519PrivateKey) Sign(data []byte) ([]byte, error) {
	return ed25519.Sign(k.Key, data), nil
}
